executor/hnswlib_searcher.py

```diff
@@ -0,0 +1,349 @@
+__copyright__ = "Copyright (c) 2021 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+import json
+from datetime import datetime
+from typing import Dict, Iterable, Optional, Generator, Tuple
+
+import hnswlib
+import numpy as np
+from bidict import bidict
+from jina import Executor, DocumentArray, Document
+from jina.logging.logger import JinaLogger
+
+GENERATOR_DELTA = Generator[
+    Tuple[str, Optional[np.ndarray], Optional[datetime]], None, None
+]
+
+HNSW_TYPE = np.float32
+DEFAULT_METRIC = 'cosine'
+
+
+class HnswlibSearcher:
+    """Hnswlib powered vector indexer.
+
+    This indexer uses the HNSW algorithm to index and search for vectors. It does not
+    require training, and can be built up incrementally.
+    """
+
+    def __init__(
+        self,
+        limit: int = 10,
+        metric: str = DEFAULT_METRIC,
+        dim: int = 0,
+        max_elements: int = 1_000_000,
+        ef_construction: int = 400,
+        ef_query: int = 50,
+        max_connection: int = 64,
+        dump_path: Optional[str] = None,
+        traversal_paths: Iterable[str] = ('r',),
+        is_distance: bool = True,
+        last_timestamp: datetime = datetime.min,
+        *args,
+        **kwargs,
+    ):
+        """
+        :param limit: Number of results to get for each query document in search
+        :param metric: Distance metric type, can be 'euclidean', 'inner_product',
+        or 'cosine'
+        :param dim: The dimensionality of vectors to index
+        :param max_elements: Maximum number of elements (vectors) to index
+        :param ef_construction: The construction time/accuracy trade-off
+        :param ef_query: The query time accuracy/speed trade-off. High is more
+            accurate but slower
+        :param max_connection: The maximum number of outgoing connections in the
+            graph (the "M" parameter)
+        :param dump_path: The path to the directory from where to load, and where to
```

Comment:
If that is true and we don't envision using the `dump_path` for actual loading of the index: Can we just remove the `dump_path` completely and rely on workspace for the space on disc where to write?

Reply:
No, we would use it, I think, instead of the workspace itself. Because it's easier to mount the volume to any location and pass that here, rather than calculate where the workspace would be in the Indexer on the machine.  The question is if we should keep it here while it's not meant to be used (for the time).